### PR TITLE
feat(ext/web): BYOB support for ReadableStream

### LINKED ISSUE #11025: Support BYOB ReadableStream

### PR DESCRIPTION
**Pull request by crowlKats**
Closes #11025

### CODE REVIEW

**Review by kitsonk — COMMENTED** [review]
Are there no WPT tests that need to be enabled?

**Comment by crowlKats** [comment]
Yes, its not completely done yet. there are a few TODOs that hopefully will get resolved in the morning. once those are solved, I will enable all the WPT tests related to this

File: ext/web/06_streams.js [hunk 1]
```diff
@@ -1019,112 +1193,629 @@
...
+   * @returns {void}
+   */
+  function readableByteStreamControllerFillHeadPullIntoDescriptor(
+    controller,
+    size,
+    pullIntoDescriptor,
+  ) {
+    assert(
+      controller[_pendingPullIntos].length === 0 ||
+        controller[_pendingPullIntos][0] === pullIntoDescriptor, // TODO: check
```
Comment (lucacasonato) [inline ext/web/06_streams.js:1638]: Looks correct. You doing a reference equality check here, not a value equality check, which looks to be what the spec intends.

File: ext/web/06_streams.js [hunk 2]
```diff
@@ -1138,7 +1829,12 @@
     assert(signal === undefined || signal instanceof AbortSignal);
     assert(!isReadableStreamLocked(source));
     assert(!isWritableStreamLocked(dest));
-    const reader = acquireReadableStreamDefaultReader(source);
+    let reader;
+    if (source[_controller] instanceof ReadableByteStreamController) {
+      reader = acquireReadableStreamBYOBReader(source);
```
Comment (lucacasonato) [inline ext/web/06_streams.js:1834]: > If source.[[controller]] implements ReadableByteStreamController, let reader be either ! AcquireReadableStreamBYOBReader(source) or ! AcquireReadableStreamDefaultReader(source), at the user agent’s discretion.

I'll have to dig into that deeper. In what cases would a user agent use a DefaultReader?

File: ext/web/06_streams.js [hunk 3]
```diff
@@ -3450,10 +4443,201 @@
...
+      return promise.promise;
+    }
+
+    /** @returns {void} */
+    releaseLock() {
+      webidl.assertBranded(this, ReadableStreamBYOBReader);
+      if (this[_stream] === undefined) {
+        return;
+      }
+      if (this[_readIntoRequests].length) {
```
Comment (lucacasonato) [inline ext/web/06_streams.js:4527]:
```suggestion
      if (this[_readIntoRequests].length !== 0) {
```

File: ext/web/06_streams.js [hunk 4]
```diff
@@ -3450,10 +4443,201 @@
...
+        throw new TypeError(
+          "There are pending read requests, so the reader cannot be release.",
```
Comment (lucacasonato) [inline ext/web/06_streams.js:4529]:
```suggestion
          "There are pending read requests, so the reader cannot be released.",
```

**Review by lucacasonato — COMMENTED** [review]
Ok, did an initial review. This looks very promising.

File: ext/web/06_streams.js [hunk 5]
```diff
@@ -193,6 +207,20 @@
     return isFakeDetached in O;
   }
 
+  /**
+   * @param {ArrayBufferLike} O
+   * @returns {boolean}
+   */
+  function canTransferArrayBuffer(O) {
+    assert(typeof O === "object");
+    assert(O instanceof ArrayBuffer || O instanceof SharedArrayBuffer);
```
Comment (andreubotella) [inline ext/web/06_streams.js:216]: `SharedArrayBuffer`s aren't transferable, and I don't think this spec algorithm is meant to be passed a SAB.

File: ext/web/06_streams.js [hunk 6]
```diff
@@ -1019,112 +1193,629 @@
...
+      view instanceof Uint8ClampedArray ||
+      view instanceof Int16Array ||
+      view instanceof Uint16Array ||
+      view instanceof Int32Array ||
+      view instanceof Uint32Array ||
+      view instanceof BigInt64Array ||
+      view instanceof BigUint64Array
+    ) {
+      elementSize = view.constructor.BYTES_PER_ELEMENT;
+      ctor = view.constructor;
```
[the remark this answers is not in the document]
Comment (crowlKats) [reply ext/web/06_streams.js:1270]: true. not sure how we can do this better though, except a check for each individual typed array

File: ext/web/06_streams.js (hunk 2 above)
Comment (crowlKats) [reply ext/web/06_streams.js:1834]: Not sure to be honest

File: ext/web/06_streams.js [hunk 7]
```diff
@@ -193,6 +207,20 @@
...
+   * @param {ArrayBufferLike} O
+   * @returns {boolean}
+   */
+  function canTransferArrayBuffer(O) {
+    assert(typeof O === "object");
+    assert(O instanceof ArrayBuffer || O instanceof SharedArrayBuffer);
+    if (isDetachedBuffer(O)) {
+      return false;
+    }
+    // TODO(@crowlKats): 4. If SameValue(O.[[ArrayBufferDetachKey]], undefined) is false, return false.
```
Comment (andreubotella) [inline ext/web/06_streams.js:220]: I believe this check is the same as [`v8::ArrayBuffer::is_detachable`](https://docs.rs/v8/0.33.0/v8/struct.ArrayBuffer.html#method.is_detachable). The use case for this seems to be to make it impossible to detach the buffer that backs a `WebAssembly.Memory` object other than through its `grow` method.

File: ext/web/06_streams.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lucacasonato) [reply ext/web/06_streams.js:1273]: It's probably fine. We can fix it if someone runs into it.

File: ext/web/06_streams.js (hunk 7 above)
Comment (lucacasonato) [reply ext/web/06_streams.js:220]: Yes. I suggest we do this in a follow up, as it requires core bindings.

File: ext/web/06_streams.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreubotella) [reply ext/web/06_streams.js:1273]: Instances of `MyArray` should still have a `[[TypedArrayName]]` internal slot of `"Int8Array"`. Now, that would not be if the constructor returns something other than `this`, or with `Object.create(MyArray.prototype)`, but that's even more obscure, so it's probably fine to just have a TODO.

File: tools/wpt/expectation.json [hunk 8]
```diff
@@ -6219,231 +6149,21 @@
...
-        "calling respondWithNewView() twice on the same byobRequest should throw",
-        "calling respond(0) twice on the same byobRequest should throw even when closed",
-        "calling respond() should throw when canceled",
-        "pull() resolving should not make releaseLock() possible",
-        "ReadableStream with byte source: default reader + autoAllocateChunkSize + byobRequest interaction",
-        "ReadableStream with byte source: respondWithNewView() with a smaller view",
-        "ReadableStream with byte source: respondWithNewView() with a zero-length view (in the closed state)",
-        "ReadableStream with byte source: respondWithNewView() with a transferred non-zero-length view (in the readable state)",
-        "ReadableStream with byte source: respondWithNewView() with a transferred zero-length view (in the closed state)"
+        "ReadableStream with byte source: Respond to multiple pull() by separate enqueue()"
```
Comment (lucacasonato) [inline tools/wpt/expectation.json:6157]: After investigation I am about 80% confident this test being broken is a result of our broken microtask queuing: https://github.com/denoland/deno/issues/11731

**Review by lucacasonato — APPROVED** [review]
Thank you @crowlKats. Awesome to have this implemented now!

Other than the one failing test due to microtask queuing, all WPT pass.

**Pull request merged**